app/screens/browse_channels/browse_channels.tsx

```diff
@@ -0,0 +1,501 @@
+                            }
+                            setLoading(false);
+
+                            break;
+                    }
+                }
+            };
+            switch (typeOfChannels) {
+                case PUBLIC:
+                    if (nextPublic.current) {
+                        setLoading(true);
+                        fetchChannels(
+                            serverUrl,
+                            currentTeamId,
+                            publicPage.current + 1,
+                            General.CHANNELS_CHUNK_SIZE,
+                        ).then(({channels: receivedChannels}) => loadedChannels(receivedChannels || [])); // Handle error?
+                    }
+                    break;
+                case SHARED:
+                    if (nextShared.current) {
+                        setLoading(true);
+                        fetchSharedChannels(
+                            serverUrl,
+                            currentTeamId,
+                            sharedPage.current + 1,
+                            General.CHANNELS_CHUNK_SIZE,
+                        ).then(({channels: receivedChannels}) => loadedChannels(receivedChannels || []));
+                    }
+                    break;
+                case ARCHIVED:
+                default:
+                    if (canShowArchivedChannels && nextArchived.current) {
+                        setLoading(true);
+                        fetchArchivedChannels(
+                            serverUrl,
+                            currentTeamId,
+                            archivedPage.current + 1,
+                            General.CHANNELS_CHUNK_SIZE,
+                        ).then(({channels: receivedChannels}) => loadedChannels(receivedChannels || []));
+                    }
+                    break;
+            }
+        }
+    }, [loading, typeOfChannels, joinedChannels, archivedChannels, channels, sharedChannels]);
+
+    const selectActiveChannels = useCallback(() => {
+        switch (typeOfChannels) {
+            case ARCHIVED:
+                return archivedChannels;
+            case SHARED:
+                return sharedChannels;
+            default:
+                return channels;
+        }
+    }, [archivedChannels, channels, sharedChannels, typeOfChannels]);
+
+    const stopSearch = useCallback(() => {
+        setVisibleChannels(selectActiveChannels());
+        setTerm('');
+    }, [selectActiveChannels]);
+
+    const searchChannels = useCallback((text: string) => {
+        if (text) {
+            const active = selectActiveChannels();
+            const filtered = filterChannelsByTerm(active, text);
+            setTerm(text);
+            if (
+                filtered.length !== visibleChannels.length ||
+                filtered.reduce((shouldUpdate, c, i) => shouldUpdate || c.id !== visibleChannels[i].id, false)
+            ) {
+                setVisibleChannels(filtered);
+            }
```

Comment:
This might be an early optiimization, but I was trying to reduce the number of times we update the visible channels and therefore the flat list.